hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java

```diff
@@ -174,48 +312,148 @@ public void bumpRejectedType3EthTx(final long bumpBy) {
         }
     }
 
+    // ---------------------------------------------
+    // P2 metrics: System contract per-method counts
+
+    enum MethodResult {
+        SUCCESS,
+        FAILURE;
+
+        public static @NonNull MethodResult from(@NonNull final MessageFrame.State state) {
+            return switch (state) {
+                case State.CODE_SUCCESS, State.COMPLETED_SUCCESS -> SUCCESS;
+                default -> FAILURE;
+            };
+        }
+    }
+
+    public void incrementSystemMethodCall(
+            @NonNull SystemContractMethod systemContractMethod, @NonNull final MessageFrame.State state) {
+        systemContractMethod =
+                systemContractMethodRegistry.fromMissingContractGetWithContract(requireNonNull(systemContractMethod));
+        requireNonNull(state);
+
+        final var result = MethodResult.from(state);
+        if (p2MetricsEnabled) {
+            // Handle each of the P2 metrics kinds
+
+            try {
+                final Consumer<Counter[]> bumpMetricsPair = metricsPair -> {
+                    metricsPair[MethodMetricType.TOTAL.index].increment();
+                    if (MethodResult.FAILURE == result) {
+                        metricsPair[MethodMetricType.FAILED.index].increment();
+                    }
+                };
+
+                bumpMetricsPair.accept(systemContractMethodCounters.get(
+                        systemContractMethod.systemContract().orElse(null)));
+                bumpMetricsPair.accept(systemContractMethodCountersVia.get(systemContractMethod.via()));
+                systemContractMethodErcMembers
+                        .get(systemContractMethod)
+                        .forEach(ercType -> bumpMetricsPair.accept(systemContractERCTypeCounters.get(ercType)));
+                systemContractMethodGroupMembers
+                        .get(systemContractMethod)
+                        .forEach(group -> bumpMetricsPair.accept(systemContractMethodGroupCounters.get(group)));
+            } catch (final NullPointerException e) {
+                // ignore: should never happen, but ... just in case ... don't fail tx because of bad
+                // metrics code
```

Comment:
Maybe log a warning here with any pertinent information just in case

Reply:
Left a `// FUTURE` comment.  Need to be careful to not spam logs with messages for _same_ system contract method.  So seems to require a _map_ of logs already emitted.